File: 2 - Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:

        #42ms beats 99,41%
        carry = 0

        result = ListNode()
        root = result

        while l1 or l2 or (carry > 0):
            sum = 0
            sum += carry
            carry = 0
            if l1:
                sum += l1.val
                l1 = l1.next
            if l2:
                sum += l2.val
                l2 = l2.next
            if sum >= 10:
                sum -= 10
                carry = 1

            result.val = sum
            
            result.next = ListNode() if l1 or l2 or (carry > 0) else None
            result = result.next
        return root

File: 2 - Add Two Numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    // 98ms beats 30,41%
    let carry = 0
    let sum = 0

    let result = new ListNode(0)
    let root = result

    while(l1 || l2 || carry > 0) {
        sum = carry
        carry = 0
        if(l1) {
            sum += l1.val
            l1 = l1.next
        }
        if(l2) {
            sum += l2.val
            l2 = l2.next

        }
        if(sum >= 10) {
            sum -= 10
            carry = 1
        }

        result.val = sum

        result.next = l1 || l2 || (carry > 0) ? new ListNode(0) : null
        result = result.next
    }

    return root
        
};
